src/libtrx/game/ui/dialogs/settings.c

```diff
@@ -0,0 +1,359 @@
+        if (dir < 0) {
+            return *(float *)option->target > (float)option->min_value / 100;
+        } else if (dir > 0) {
+            return *(float *)option->target < (float)option->max_value / 100;
+        }
+        break;
+    case COT_RGB888: {
+        const uint8_t *const component = M_GetColorComponent(option);
+        if (dir < 0) {
+            return *component > option->min_value;
+        } else if (dir > 0) {
+            return *component < option->max_value;
+        }
+        break;
+    }
+    case COT_ENUM: {
+        const M_ENUM_LOOKUP enum_lookup = M_GetEnumEntry(option);
+        ASSERT(enum_lookup.entry != nullptr);
+        if (dir < 0) {
+            return enum_lookup.position > 0;
+        } else if (dir > 0) {
+            return enum_lookup.position < enum_lookup.count - 1;
+        }
+        break;
+    }
+    default:
+        break;
+    }
+    return false;
+}
+
+static bool M_RequestChangeValue(
+    const UI_SETTINGS_STATE *const s, const int32_t row_idx, const int32_t dir)
+{
+    if (!M_CanChangeValue(s, row_idx, dir)) {
+        return false;
+    }
+
+    const UI_SETTINGS_OPTION *const option = &s->options[row_idx];
+    int32_t delta = g_Input.slow ? option->delta_slow : option->delta_fast;
+    if (delta == 0) {
+        delta = 1;
+    }
+    delta *= dir;
+
+    if (option->custom_handler.request_change_value != nullptr) {
+        if (option->custom_handler.request_change_value(option, dir)) {
+            goto changed;
+        }
+        return false;
+    }
+
+    switch (option->option_type) {
+    case COT_BOOL:
+        *(bool *)option->target = !*(bool *)option->target;
+        break;
+    case COT_INT32:
+        *(int32_t *)option->target += delta;
+        break;
+    case COT_DOUBLE:
+        *(double *)option->target =
+            (round(*(double *)option->target * 10) + (delta / 10.0f)) / 10.0f;
+        break;
+    case COT_FLOAT:
+        *(float *)option->target =
+            (round(*(float *)option->target * 10) + (delta / 10.0f)) / 10.0f;
+        break;
+    case COT_RGB888: {
+        uint8_t *const component = M_GetColorComponent(option);
+        int32_t component_i = *component;
+        component_i += delta;
+        CLAMP(component_i, 0, 255);
+        *component = component_i;
+        break;
+    }
+    case COT_ENUM: {
+        const M_ENUM_LOOKUP enum_lookup = M_GetEnumEntry(option);
+        const UI_SETTINGS_ENUM_ENTRY *const next_entry =
+            &((UI_SETTINGS_ENUM_ENTRY *)
+                  option->misc)[enum_lookup.position + delta];
+        *(int32_t *)option->target = next_entry->value;
+        break;
+    }
+    default:
+        return false;
+    }
+
+changed:
+    Config_Write();
+    return true;
+}
+
+static float M_GetValueWidth(const UI_SETTINGS_STATE *const s)
+{
+    // Measure the maximum width of the value label to prevent the entire
+    // dialog from changing its size as the player changes the sound levels.
```

Comment:
Really minor, but the comment refers to sound levels.

Reply:
Thanks – fixed